File: windows-app/src/ClipboardSync.App/Clipboard/LoopGuard.cs
namespace ClipboardSync.App.Clipboard;

public sealed class LoopGuard
{
    private readonly int _maxEntries;
    private readonly TimeSpan _suppressionWindow;
    private readonly Dictionary<string, DateTimeOffset> _eventIds = new();
    private readonly Dictionary<string, DateTimeOffset> _suppressedHashes = new();

    public LoopGuard(int maxEntries = 128, TimeSpan? suppressionWindow = null)
    {
        _maxEntries = maxEntries;
        _suppressionWindow = suppressionWindow ?? TimeSpan.FromSeconds(4);
    }

    public void RememberSeenEvent(string eventId)
    {
        _eventIds[eventId] = DateTimeOffset.UtcNow;
        Trim(_eventIds);
    }

    public bool HasSeenEvent(string eventId) => _eventIds.ContainsKey(eventId);

    public void MarkRemoteApplied(string hash)
    {
        _suppressedHashes[hash] = DateTimeOffset.UtcNow.Add(_suppressionWindow);
        Trim(_suppressedHashes);
    }

    public bool ShouldSuppressLocal(string hash)
    {
        if (!_suppressedHashes.TryGetValue(hash, out var until))
        {
            return false;
        }

        if (until < DateTimeOffset.UtcNow)
        {
            _suppressedHashes.Remove(hash);
            return false;
        }

        return true;
    }

    private void Trim(Dictionary<string, DateTimeOffset> dictionary)
    {
        if (dictionary.Count <= _maxEntries)
        {
            return;
        }

        foreach (var key in dictionary.OrderBy(pair => pair.Value).Take(dictionary.Count - _maxEntries).Select(pair => pair.Key).ToList())
        {
            dictionary.Remove(key);
        }
    }
}

File: android-app/app/src/main/java/com/clipboardsync/android/clipboard/LoopGuard.kt
package com.clipboardsync.android.clipboard

import java.util.LinkedHashMap

class LoopGuard(
    private val maxEntries: Int = 128,
    private val suppressionWindowMillis: Long = 4_000L
) {
    private val recentEventIds = object : LinkedHashMap<String, Long>(maxEntries, 0.75f, true) {
        override fun removeEldestEntry(eldest: MutableMap.MutableEntry<String, Long>?): Boolean = size > maxEntries
    }
    private val suppressedHashes = object : LinkedHashMap<String, Long>(maxEntries, 0.75f, true) {
        override fun removeEldestEntry(eldest: MutableMap.MutableEntry<String, Long>?): Boolean = size > maxEntries
    }

    fun rememberSeenEvent(eventId: String) {
        recentEventIds[eventId] = System.currentTimeMillis()
    }

    fun hasSeenEvent(eventId: String): Boolean = recentEventIds.containsKey(eventId)

    fun markRemoteApplied(hash: String) {
        suppressedHashes[hash] = System.currentTimeMillis() + suppressionWindowMillis
    }

    fun shouldSuppressLocal(hash: String): Boolean {
        val until = suppressedHashes[hash] ?: return false
        if (until < System.currentTimeMillis()) {
            suppressedHashes.remove(hash)
            return false
        }
        return true
    }
}
